Mathlib/Analysis/SpecialFunctions/BinEntropy.lean

```diff
@@ -0,0 +1,748 @@
+    rw [this]
+    exact logConcave
+
+lemma binaryEntropy_lt_log2_of_gt_half {p : ℝ} : 1/2 < p → p ≤ 1 → binaryEntropy p < log 2 := by
+  intros
+  rw [← binaryEntropy_one_sub]
+  exact binaryEntropy_lt_log2_of_lt_half (by linarith) (by linarith)
+
+lemma binaryEntropy_eq_log2_iff_eq_half {p : ℝ} (pge0 : 0 ≤ p) (ple1 : p ≤ 1) :
+    binaryEntropy p = log 2 ↔ p = 1/2 := by
+  constructor <;> intro h
+  · by_cases h' : p < 1/2
+    · linarith [binaryEntropy_lt_log2_of_lt_half pge0 h']
+    · by_cases pgthalf : 1/2 < p
+      · have := binaryEntropy_lt_log2_of_gt_half pgthalf ple1
+        linarith
+      · linarith
+  · simp only [one_div, binaryEntropy_onehalf, h]
+
+lemma binaryEntropy_le_log2 {p : ℝ} (pge0 : 0 ≤ p) (ple1 : p ≤ 1) :
+    binaryEntropy p ≤ log 2 := by
+  by_cases hh: p = 1/2
+  · simp only [one_div, binaryEntropy_onehalf, le_refl, hh]
+  · by_cases gg: binaryEntropy p = log 2
+    · simp only [le_refl, gg]
+    · by_cases hhh: p < 1/2
+      · linarith [binaryEntropy_lt_log2_of_lt_half pge0 hhh]
+      · have : 1/2 < p := by
+          refine Ne.lt_of_le ?_ ?_
+          · simp only [not_lt] at hhh
+            intro a
+            simp_all only [not_true_eq_false]
+          · simp_all only [one_div, not_lt]
+        have := binaryEntropy_lt_log2_of_gt_half this ple1
+        exact LT.lt.le this
+
+/-- The q-ary entropy function is continuous everywhere.
+This is due to definition of `Real.log` for negative numbers. -/
+@[fun_prop] lemma qaryEntropy_continuous {q : ℕ} : Continuous (qaryEntropy q) := by
+  refine Continuous.add ?_ (Continuous.neg ?_)
+  · exact Continuous.sub (by fun_prop) continuous_mul_log
+  · exact Continuous.comp continuous_mul_log (continuous_sub_left 1)
+
+/-- Binary entropy is continuous everywhere.
+This is due to definition of `Real.log` for negative numbers. -/
+@[fun_prop] lemma binaryEntropy_continuous : Continuous binaryEntropy := qaryEntropy_continuous
+
+/-! ### Derivatives of binary entropy function -/
+
+@[simp] lemma deriv_one_minus (x : ℝ) : deriv (fun (y : ℝ) ↦ 1 - y) x = -1 := by
+  have onem (y : ℝ) : 1 - y = -(y + -1) := by ring
+  simp_rw [onem]
+  simp only [neg_add_rev, neg_neg, differentiableAt_const, deriv_const_add', deriv_neg'']
+
+@[simp] lemma differentiable_const_minus {q : ℝ} (p : ℝ) :
+    DifferentiableAt ℝ (fun p => q - p) p := by
+  have (p : ℝ) : q - p = -(p - q) := by ring
+  simp_rw [this]
+  apply differentiableAt_neg_iff.mpr
+  apply DifferentiableAt.add_const
+  exact differentiableAt_id'
+
+section general
+
+variable {𝕜 : Type*} [NontriviallyNormedField 𝕜]
+variable {E : Type*} [NormedAddCommGroup E] [NormedSpace 𝕜 E]
+variable {F : Type*} [NormedAddCommGroup F] [NormedSpace 𝕜 F]
+variable {f f₀ f₁ g : E → F}
+variable {x : E}
+
+-- TODO clean up, put somewhere else?
+lemma not_DifferentiableAt_of_not_DifferentiableAt_add_DifferentiableAt_
+    (hf : ¬ DifferentiableAt 𝕜 f x) (hg : DifferentiableAt 𝕜 g x) :
+    ¬ DifferentiableAt 𝕜 (fun y => f y + g y) x := by
```

Comment:
```suggestion
lemma DifferentiableAt.add_iff_left (hg : DifferentiableAt 𝕜 g x) :
    DifferentiableAt 𝕜 (fun y => f y + g y) x ↔ DifferentiableAt 𝕜 f x := by
```
and this will need to move to an earlier file